climakitae/threshold_tools.py

```diff
@@ -14,41 +14,180 @@
+            da_series = da_series.resample(time=f"{group_len}D", label="left").max()
+        elif extremes_type == "min":
+            da_series = da_series.resample(time=f"{group_len}D", label="left").min()
+
+    if grouped_duration != None:
+        if groupby == None:
+            raise ValueError(
+                "To use `grouped_duration` option, must first use groupby."
+            )
+        # In this case, identify the min (max) value of the grouped values for
+        # each window of length `grouped_duration``. Must be in `days`.
+        dur2_len, dur2_type = grouped_duration
+        if dur2_type != "day":
+            raise ValueError(
+                "`grouped_duration` specification must be in days. example: `grouped_duration = (3, 'day')`."
+            )
+
+        # Now select the min (max) from the duration period
+        if extremes_type == "max":
+            da_series = da_series.rolling(time=dur2_len, center=False).min("time")
+        elif extremes_type == "min":
+            da_series = da_series.rolling(time=dur2_len, center=False).max("time")
+
+    # Now select the most extreme value for each block in the series
     if extremes_type == "max":
-        ams = da.resample(time="A").max(keep_attrs=True)
-        ams.attrs["extreme value extraction method"] = "block maxima"
-        ams.attrs["extremes type"] = "maxima"
-        ams.attrs["block size"] = "1 year"
-        ams.attrs["timeseries type"] = "annual max series"
+        bms = da_series.resample(time=f"{block_size}A").max(keep_attrs=True)
+        bms.attrs["extremes type"] = "maxima"
+    elif extremes_type == "min":
+        bms = da_series.resample(time=f"{block_size}A").min(keep_attrs=True)
+        bms.attrs["extremes type"] = "minima"
+
+    # Calculate the effective sample size of the computed event type in all blocks, check the average value
+    if check_ess:
+        all_ess = []
+
+        # handle the ESS check differently depending on if it's spatial data
+        if "x" in da_series.dims and "y" in da_series.dims:
+            # case for spatial data, using stack/unstack to apply calculate_ess
+            stacked_da = (
+                da_series.stack(allpoints=["y", "x"]).dropna(dim="allpoints").squeeze()
+            )
+            for yr in set(bms.time.dt.year.values[0:-1]):
+                ess = xr.apply_ufunc(
+                    calculate_ess,
+                    stacked_da.sel(time=slice(f"{yr}", f"{yr+block_size-1}")).groupby(
+                        "allpoints"
+                    ),
+                    input_core_dims=[["time"]],
+                ).unstack("allpoints")
+                all_ess.append(ess)
+        else:
+            # case for not spacial data (no x, y dimensions)
+            for yr in set(bms.time.dt.year.values[0:-1]):
+                ess = calculate_ess(
+                    da_series.sel(time=slice(f"{yr}", f"{yr+block_size-1}"))
+                )
+                all_ess.append(ess)
```

Comment:
Looking for feedback on how to re-write this section better (cleaner, more efficient). Just trying to apply the `calculate_ess` function to each dimension (x, y) except for time, which it has to be applied to in separate year (or block_size) chunks (currently as a for loop). I tried to group by year, then call apply, but couldn't get it to work. The current setup uses stack/unstack for the case of x,y data, but am wondering if there is a simpler way to do this.